### ISSUE TITLE
Performance + power consumption issue

### ISSUE THREAD

**Issue by jminor** [OpenTimelineIO/raven#37]
Raven naively uses Dear ImGui in a way that re-renders the window even when not needed. The application's performance & power consumption could be improved by fixing this oversight.

When I initially wrote Raven, this continuous rendering was problematic, and I attempted to use a branches of ImGui with support for `ImGuiConfigFlags_EnablePowerSavingMode` (see `POWER_SAVING` in the code). If I recall correctly Raven's WASM/Emscripten build suffered from this issue quite badly, and I was disappointed to see that the power saving mode didn't work in the browser environment due to specifics of how Emscripten works.

At the time, I set that aside for "later" figuring that Dear ImGui would improve and Raven's energy usage would improve by virtue of that change. As of this writing, that feature has not landed in ImGui.

Today (Feb 2024) I was curious to measure Raven's performance on macOS. A quick look at Raven in Activity Monitor shows that this is not a problem. The app consumes 0% CPU when it is not active, or when the user's cursor is not moving, and only about 10% when the user actively zooms/pans. I'm not entirely sure why or how that changed, so some investigation should be done. Also we should compare on other platforms (Windows, Linux) to see if those have also improved.

I also found that ocornut, Dear ImGui's author, has written this advice on the subject which looks pretty straight forward to implement, and which proposes a solution for the Emscripten scenario: https://github.com/ocornut/imgui/wiki/Implementing-Power-Save,-aka-Idling-outside-of-ImGui

So this issue is here to ask for help with the following:
- Measure Raven's CPU usage when active vs inactive on these platforms:
  - Windows
  - Linux
  - Emscripten
- *If* those are problematic, then attempt to implement the strategy outlined in the linked wiki page above.

**Comment by jminor** [comment]
Here's a simple CPU 0% measurement when idle on macOS (M1 MacBook Pro, running macOS 14.3):
<img width="1010" alt="Screenshot 2024-02-02 at 12 24 22 AM" src="https://github.com/OpenTimelineIO/raven/assets/169599/fecf1bb0-263c-49f1-be7f-76f3de89eaf3">

and CPU 15% when active:
<img width="1010" alt="Screenshot 2024-02-02 at 12 25 04 AM" src="https://github.com/OpenTimelineIO/raven/assets/169599/0e333f08-cf5a-46bf-b49b-7418d3c819a5">

**Comment by darbyjohnston** [comment]
On Windows 10 compiled in "Release" mode with the AWS Picchu .otio file:

* Idle: ~0.5%-1.0% CPU and ~2%-5% GPU, power consumption low to very low
* Panning: ~0.5%-1.0% CPU and ~2%-6% GPU, power consumption low to very low

The numbers are from the Windows Task Manager.

For comparison the tlRender example applications (both GLFW and Qt based) have about the same performance when panning but go down to 0% when idle.

**Comment by darbyjohnston** [comment]
On Ubuntu 22.04 compiled in "Release" mode with the AWS Picchu .otio file:
* Idle: ~25%-27% CPU
* Panning: ~33% CPU

The numbers are from the System Monitor.

That's strange, I wonder why there is such a large difference on Linux.

**Comment by jminor** [comment]
Thanks for testing this on the other platforms @darbyjohnston ! It looks like this is a bigger issue on Linux than the other platforms. I wonder if some operating systems throttle apps that aren't getting user input?

**Comment by jminor** [comment]
Interesting analysis of immediate mode UIs: https://www.forrestthewoods.com/blog/proving-immediate-mode-guis-are-performant/

**Comment by meshula** [comment]
checking a mac build just now, swirling the mouse vigorously, it jumps up to a full core, but rapidly drops to 0 when I stop.

**Comment by jminor** [comment]
Re-reading the code, I see that there is a call to `glfwWaitEvents()` in the macOS main loop which is why CPU drops to zero when no mouse events are happening. The Linux main loop has a similar line, but the Windows and Emscripten main loops do not. (see links below)

I recall now that when I was experimenting with the `#define POWER_SAVING` toggle I found that Raven does not use any animation, so it could wait for events instead of polling, thus preventing the majority of wasted energy at the cost of some slight artifacts. For example, you'll sometimes notice that clicking something in Raven does not respond until you move your mouse because it is waiting for an event (the mouse movement).

The platform differences I'm seeing in the code don't completely explain the CPU measurements earlier in this thread, but it does give us some explanation of why they are different on each platform.

For now, I think it would be sufficient to add a similar wait-for-event call to the Windows and Emscripten main loops and then re-measure to see if that improves CPU usage when the app is idle.

Mac:
https://github.com/OpenTimelineIO/raven/blob/1e57ae18cf7432f78d8a40e97d6dff12665c1794/main_macos.mm#L126

Linux:
https://github.com/OpenTimelineIO/raven/blob/1e57ae18cf7432f78d8a40e97d6dff12665c1794/main_glfw.cpp#L182

Windows:
https://github.com/OpenTimelineIO/raven/blob/1e57ae18cf7432f78d8a40e97d6dff12665c1794/main_win32.cpp#L36

Emscripten:
https://github.com/OpenTimelineIO/raven/blob/1e57ae18cf7432f78d8a40e97d6dff12665c1794/main_emscripten.cpp#L133

**Comment by meshula** [comment]
Oh, I was wondering where the "jiggle the mouse to respond" was coming from. An easy remedy for that is to put a count down on the call to glfwWaitEvent. Trivially, an offending MenuItem response could set the countdown to "1" giving one period of update to avoid the jiggle. Ideally, there'd be a way to do that (presumably inside imgui) so that everything just gets the behavior. Perhaps in the platform backends under our control, we add the countdown to all mouse up events, and have the mainloop query the backend for what we might call a detected idle state. If the main viewport was refreshing due to animation, the animating mechanism itself could tell the backend "we're not idle this frame".

**Comment by jminor** [comment]
Yeah, a countdown or minimum number of redraws after an event will cover all cases I can think of. Probably just assuming 2 or 3 redraws would cover it. The advice from Dear ImGui's author linked at the top of this thread offers some specific info about how to accomplish this with Emscripten which doesn't have a traditional main loop.

Would one of y'all with a Windows dev environment mind trying out a wait-for-event in the Windows main loop?

**Comment by ThomasWilshaw** [comment]
I've tried adding `WaitMessage()` to `MessageFibreProcess()` in `main_win32.cpp` and CPU usage falls to 0% after a short while of inactivity. I've also not noticed any of the glitches you mentioned but that might just be me.

**Comment by ThomasWilshaw** [comment]
To add to this, I was looking at the in built metrics viewer (View->Metrics) and noticed that when I stop moving the mouse it consitently keeps updating for 2-8 seconds before the UI freezes. It seems a bit random how long it keeps updating for but maybe that's why I've not noticed any glitching as it always takes at least a couple of seconds to stop updating which is more than enough time to run any operations.

**Comment by meshula** [comment]
Quick test:

```
#else
            static int cooldown = 20;
            if (cooldown <= 0) {
                glfwWaitEvents();
                cooldown = 20;
            }
            else
                --cooldown;
#endif
```
gives time for imgui to do "all the things" during start up and then allows "all the things" the app wants to do after an interaction to occur before idling.

I haven't played with the number, and can only report that a cooldown of 2 is not enough.

**Comment by jminor** [comment]
That looks like a great strategy @meshula !

FYI, I also noticed that [ImHex](https://imhex.werwolv.net) seems to do a good job of ramping down CPU when inactive. It might be worth looking to see what their strategy is. That's the most mature Dear ImGui app I know of.

Also both ImHex and [Tracy](https://github.com/wolfpld/tracy) run in the browser, so they might have a good Emscripten strategy in place.

**Comment by ThomasWilshaw** [comment]
For reference ImHex's main loop is [here](https://github.com/WerWolv/ImHex/blob/2b715bb0deccefdbd7545bbd6a8e5b57b4fa0bea/main/gui/source/window/window.cpp#L212) and they do indeed have a stratergy for this. It looks like they have a sleep mode that renders at 5fps and then various callbacks that trigger a higher fps render rate. They also use glfwWaitEvent if the app is minimised or otherwise not visible. However as far as I can tell they run everything through glfw and aren't having to support multiple back ends so it's easier to maintain a more complex solution.

**Comment by jminor** [comment]
Tracy uses a simpler strategy [here](https://github.com/wolfpld/tracy/blob/4a5a21cdb08c3b554bf1857b4f2f33d7f7db0207/profiler/src/BackendGlfw.cpp#L140). It just sleeps 50ms between loops when the window doesn't have focus. That doesn't seem ideal to me.

**Comment by jminor** [comment]
I've implemented @meshula's suggestion from above here: https://github.com/OpenTimelineIO/raven/pull/99

**Comment by ThomasWilshaw** [comment]
I've added a PR #102 that hopefully brings Windows in line with these changes too

**Comment by jminor** [comment]
Thanks for #102 @ThomasWilshaw! That addresses the issue on Windows.

Now we're only left with WASM to address. I've updated the title of this issue to reflect that.
